### PR TITLE
Shortcode templates

### PR DESCRIPTION
**Pull request by deanmarcussen**
Fixes https://github.com/OrchardCMS/OrchardCore/issues/6199

wip. so not fully implemented.

Includes shortcode popover with selector, and shortcode liquid templates, edited from the admin.

Refer tonights meeting for demos and visuals.
Also fixes https://github.com/OrchardCMS/OrchardCore/issues/6199

TODO
- [x] Test Fluid html encoding shortcode templates
- [x] Add Field Definitions to context
- [x] Implement modal
- [x] Keep categories
- [x] Lower case shortcode name
- [x] Hint in template ui
- [x] Include in starter templates

Docs
- [x] Document how to sanitize a shortcode template in liquid
- [x] Upgrade from rc2 note

### CODE REVIEW

**Comment by agriffard** [comment]
An error 'Shape ShortcodePopover not found' occurred because the Shortcodes feature was not enabled.
Of course, enabling the feature solves the problem.

It will happen to most people using RC2 and migrating to next version having this Shortcode module.

**Comment by deanmarcussen** [comment]
UI Tweaks (you got your preview Seb as it was a nice feature to add) and even some docs

Screenshot 2020-08-01 at 15 22 40

File: src/docs/reference/modules/Shortcodes/README.md [hunk 1]
```diff
@@ -1,11 +1,137 @@
...
+
+### `IShortcodeProvider`
+
+Shortcodes may also be used by implementing the `IShortcodeProvider` interface and registered with the `AddShortcode` extension method.
+
+In this example we register an `ImageShortcodeProvider` as `[image]` and describe the Shortcode.
+
+``` csharp
+services.AddShortcode<ImageShortcodeProvider>("image", describe => {
+    describe.ReturnShortcode = "[image] [/image]";
```
Comment (sebastienros) [inline src/docs/reference/modules/Shortcodes/README.md:109]: DefaultValue ?

File: src/docs/reference/modules/Shortcodes/README.md [hunk 2]
```diff
@@ -1,11 +1,137 @@
...
+    describe.Hint = (serviceProvider) => {
```
Comment (sebastienros) [inline src/docs/reference/modules/Shortcodes/README.md:110]: No localization here, will use dynamic localization when available. Even though this is registered by code. We are following what we do with content types.

File: src/OrchardCore.Modules/OrchardCore.Shortcodes/Services/ShortcodeOptionsTableProvider.cs [hunk 3]
```diff
@@ -0,0 +1,35 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.Extensions.Options;
+using System.Threading.Tasks;
+
+namespace OrchardCore.Shortcodes.Services
+{
+    public class ShortcodeOptionsTableProvider : IShortcodeTableProvider
```
Comment (sebastienros) [inline src/OrchardCore.Modules/OrchardCore.Shortcodes/Services/ShortcodeOptionsTableProvider.cs:9]: Descriptor instead of Table

Comment (sebastienros) [inline src/OrchardCore.Modules/OrchardCore.Shortcodes/Services/ShortcodeOptionsTableProvider.cs:9]: This class returns ShortcodeDescriptor instances registered in the IOption<foo>.

File: src/OrchardCore/OrchardCore.DisplayManagement.Liquid/OrchardCoreBuilderExtensions.cs [hunk 4]
```diff
@@ -39,6 +39,7 @@ public static OrchardCoreBuilder AddLiquidViews(this OrchardCoreBuilder builder)
 
                 services.AddLiquidFilter<AppendVersionFilter>("append_version");
                 services.AddLiquidFilter<ResourceUrlFilter>("resource_url");
+                services.AddLiquidFilter<SanitizeFilter>("sanitize");
```
Comment (sebastienros) [inline src/OrchardCore/OrchardCore.DisplayManagement.Liquid/OrchardCoreBuilderExtensions.cs:42]: sanitize_html

Comment (deanmarcussen) [reply src/OrchardCore/OrchardCore.DisplayManagement.Liquid/OrchardCoreBuilderExtensions.cs]: All done

File: src/OrchardCore/OrchardCore.Shortcodes.Abstractions/ServiceCollectionExtensions.cs [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+{
+    public static class ServiceCollectionExtensions
+    {
+        public static IServiceCollection AddShortcode<T>(this IServiceCollection services, string name) where T : class, IShortcodeProvider =>
+            services.AddShortcode(name, null);
+
+        public static IServiceCollection AddShortcode<T>(this IServiceCollection services, string name, Action<ShortcodeOption> describe) where T : class, IShortcodeProvider
+        {
+            services.Configure<ShortcodeOptions>(options => options.AddShortcode(name, describe));
+            services.AddScoped<T>();
```
Comment (KshitizGIT) [inline src/OrchardCore/OrchardCore.Shortcodes.Abstractions/ServiceCollectionExtensions.cs:15]: Hi @deanmarcussen ,

This is breaking the feature of image short code. After this, ImageShortcodeProvider is never registered as a  IShortcodeProvider in the DI. This should be changed to 
```csharp
services.AddScope<IShortcodeProvider,T>();

```

**Pull request merged**